### ISSUE TITLE
microdom.py mishandling namespaces

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3672]
|<img alt="shywolf9982's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| shywolf9982 reported|
|-|-|
|Trac ID|trac#3672|
|Type|enhancement|
|Created|2009-02-23 17:01:08Z|
|Branch|https://github.com/twisted/twisted/tree/microdom-ns-3672|

I noticed that in [tags/releases/twisted-8.2.0/twisted/web/microdom.py,](tags/releases/twisted-8.2.0/twisted/web/microdom.py,) XML namespaces do not seem to be handled correctly.

In particular:

 * Namespaces are not propagated by default (the default scoping is not supported unless, for each
   element we add, we re-specify its namespace)

 * When rendering with prefixes (either passing the nsprefixes to the `writexml()` or by issuing
   an `addPrefixes()` on the root node) the sub-elements do not get the right prefix (actually,
   they don't get the prefix at all). This clashes a bit as per namespace scoping
   (http://www.w3.org/TR/REC-xml-names/#scoping)

 * Prefixed tag do not have the prefix on the closure

I have attached a small python script that describes the problem better than this summary does: 
what it tries to achieve is render the following XML

```
<event xmlns="http://example.com/ns/stream/event"
       xmlns:entity="http://example.com/ns/stream/entity">
  <entity:entity event-attr="on_entity">
    <entity:name><![CDATA[John Wayne]]></entity:name>
  </entity:entity>
</event>
```

and instead gets this:

```
<event xmlns="http://example.com/ns/stream/event"
       xmlns:entity="http://example.com/ns/stream/entity">
  <entity:entity event-attr="on_entity">
    <name><![CDATA[John Wayne]]></name>
  </entity>
</event>
```

Also attached is a patch (_microdom.diff_)against this file [tags/releases/twisted-8.2.0/twisted/web/microdom.py](tags/releases/twisted-8.2.0/twisted/web/microdom.py) that hopefully solves the problem.

It also adds another feature, which I'm not too sure if it is ok to put in or not, namely auto-setting of the child element namespace if this is None to the value of the parent `Element` namespace.

This works only for `Element` and subclasses, and is implemented through overriding `appendChild()`, `insertBefore()` and `replaceChild()`.

Attachments:

* [domtest.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/feb/feb8655e7d4d6aa7b95807772b8a714604c12089/7950ce919c773e3b8ef8ad735476d58ea45db1c2.py) (1292 bytes) - added by shywolf9982 on 2009-02-23 17:01:56Z - The "testing" script
* [microdom.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/feb/feb8655e7d4d6aa7b95807772b8a714604c12089/d4728c68bfe8648bb6cf59b63f322b9fca29e8dd.diff) (10903 bytes) - added by shywolf9982 on 2009-02-24 12:04:34Z - The correct patch, with the full blown tests

<details><summary>Searchable metadata</summary>

```
trac-id__3672 3672
type__enhancement enhancement
reporter__shywolf9982 shywolf9982
priority__low low
milestone__ 
branch__branches_microdom_ns_3672 branches/microdom-ns-3672
branch_author__shywolf9982__exarkun shywolf9982, exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__microdom microdom
time__1235408468000000 1235408468000000
changetime__1237480450000000 1237480450000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to shywolf9982|
|-|-|

Please write a unit test which demonstrates the problem.  domtest.py may be close, but the results need to be automatically verifiable.  See twisted/web/test/ for many examples of how to do this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Hi!  Thanks a lot for your patch, and the detailed bug report.

You should read the [wiki page on contributing to Twisted development](http://twistedmatrix.com/trac/wiki/TwistedDevelopment) to see how this process usually works.  In particular:

  * If you want someone to look at a patch you've contributed, please make sure the "review" keyword is present on the ticket.
  * Thanks for including tests of some kind - we do get a lot of contributions without tests at all :) - but your "domtest.py" script should be in the form of unit tests, with docstrings, that explain what properties are being verified.  It would be helpful for you to include in the patch rather than as a separate file (although that's not entirely necessary if it's difficult).

**Comment by twisted-trac** [comment]
|<img alt="shywolf9982's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|shywolf9982 commented|
|-|-|

Replying to [glyph](#note_2):

> You should read the [wiki page on contributing to Twisted development](http://twistedmatrix.com/trac/wiki/TwistedDevelopment) to see how this process usually works.  In particular:

Thanks for the pointers, will change the patches to comply ASAP.

>   * Thanks for including tests of some kind - we do get a lot of contributions without tests at all :) - but your "domtest.py" script should be in the form of unit tests, with docstrings, that explain what properties are being verified.  It would be helpful for you to include in the patch rather than as a separate file (although that's not entirely necessary if it's difficult).

Are doctests okay?

Anyway, will upload a correct patch ASAP, and sorry again for the inconvenience.

**Comment by twisted-trac** [comment]
|<img alt="shywolf9982's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|shywolf9982 commented|
|-|-|

Okay, I have uploaded the correct patch.

I hope I haven't done any harm (aside that poor blank line in test_xml) but all tests for microdom and twisted.web run successfully.

I'm not very satisfied with a certain thing (marked BBB in the patch) because it basically checks that the parent is not a superclass of itself (okay, it checks it's not a document).

Also, it does too much namespace magic for my liking.

However, has the plus of working, without that, the test_namespaceDelete fails miserably. And probably if that test fails, many applications might fail aswell, so I decided to put it in.

As last, since I'm using twisted (as in actually writing code that uses it and reading the documentation) since a week only, I would like to call for a close review of my code (better safe than sorry).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

The patch looks incomplete.  There's _only_ a whitespace change to `test_xml.py`, no new tests.  Is that the patch you meant to submit?

**Comment by twisted-trac** [comment]
|<img alt="shywolf9982's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|shywolf9982 commented|
|-|-|

No, the tests were in `test_microdom.py`, I forgot to svn add (duh). Now it is correct.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

That's better, thanks. :)

  1. `twisted/web/test/test_microdom.py`
    1. `twisted/web/test/test_xml.py` is already a test module with lots of tests for `twisted.web.microdom`.  New tests should probably just be added there, rather than creating a new test module.
    1. There shouldn't be a `MicroDOMTests`.  Notice in test_xml.py how the tests are roughly split up by which API they're testing, usually with one `TestCase` subclass per implementation class being tested.  This is a good rule of thumb (but it isn't the **required** test organization).
    1. The docstrings in the code are all formatted in disagreement with the coding standard.  http://twistedmatrix.com/trac/browser/trunk/doc/development/policy/coding-standard.xhtml?format=raw includes a section about docstrings and gives an example of the preferred formatting.
    1. It looks like you may be using restructured text in the docstrings.  Please use epytext markup instead.  You can find the documentation for epytext at <http://epydoc.sourceforge.net/epytext.html>.  You can also see lots of examples of epytext throughout Twisted.
    1. Test method docstrings should be of the form "X does Y" or "X is Y".  For example, I would write this for `test_prefixedTagClose`: _Elements with an XML namespace are serialized with an opening tag of C{"<namespace:tag name>"} and with a closing tag of C{"</namespace:tag name>"}._
    1. Methods should be separated from each other by two blank lines.  Classes should be separated by three blank lines.
    1. The tests look like they can be simplified somewhat.  The text nodes don't seem to be relevant to the functionality being tested.
    1. \-style line continuation is against the coding standard as well.  Use parenthesis to form a multiline string.
    1. Also, you can omit the +.  `"foo" "bar"` is the same as `"foobar"` in Python.
    1. Also, it's convenient to use `'`-style strings to contain XML, since that means you don't have to escape the `"`-style strings inside the XML.  For example, `'"'` is simpler than `"\""`.
    1. Please use `assertEquals` rather than `assertEqual` (sorry if you saw some code in Twisted that used `assertEqual`, we only recently came to a consensus on this one ;)
  1. `twisted/web/microdom.py`
    1. Twisted uses `camelCase` names for variables, not `under_scores`.  `c_endTagName` shouldn't have an underscore.  I'm not really sure what name to suggest instead though, because I'm not sure what the _c_ there even means.  Is it "convenience"?  If so, that's pointless.  You can just name the local `endTagName`.  The comment should point out _why_ it's convenient, rather than that it is.  If it's really convenient, readers will notice. :)  It's more useful to explain that it may be modified later to include a namespace prefix.
    1. Operators should be surrounded by a space.  For example, `x + ':' + y`, not `x+':'+y`.
    1. The unit tests don't cover both of the end-tag fixes.  They only cover the one on line 632, not the one on line 635. 
    1. The way top-level namespaces are implemented is indeed quite messy.  Perhaps we can come up with a better implementation strategy.  First, though, I think that feature needs tests.  The part of it which causes `xmlns=""` to not be emitted is tested, but the part that restricts that behavior to `Document` nodes or nodes without parents is not.
    1. Wherever possible, it would be good to include references to the part of the XML specification which defines the behavior being implemented.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="shywolf9982's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|shywolf9982 commented|
|-|-|

Thank you for the review :). I have fixed everything (ok, I think I did) in accordance to your suggestions and to the guidelines.

There is only a thing that I'm doubtful about

Replying to [exarkun](#note_8):

>     1. The way top-level namespaces are implemented is indeed quite messy.  Perhaps we can come up with a better implementation strategy.  First, though, I think that feature needs tests.  The part of it which causes `xmlns=""` to not be emitted is tested, but the part that restricts that behavior to `Document` nodes or nodes without parents is not.

Okay, a little bit of background here. That part was inserted because `testNamespaceDelete` fails miserably without.

So, as far as I know, the problem is the following:

 1. Upon calling of `writexml` a namespace is passed, defaulting to `''`. This can
    override the root element `xmlns`

 1. Before my patch, the recursive calls of `writexml` got the namespace of the 
    current tag being processed. Which means that inner tag in the same namespace do
    not get prefixed and this is wrong according to
    http://www.w3.org/TR/xml-names/#scoping. Unless I got it all wrong :)

 1. Hence, my patch keeps passing down the namespace without change, but this incurs
    in the following drawback: if the variable namespace is empty (`''`), but the
    root element `xmlns` is not, I have to pass down the root element namespace
    instead of the variable passed to the function. Hence why that if.

So, in short, I'm pretty sure that `testNamespaceDelete` tests that code, and a 
quick run with the coverage switch on seems to indicate it is not the only one.

I will try to come up with a decent test for it, but I'm going to check carefully if there aren't corner cases I haven't seen: sadly that code is way more complex than it looks, is quite misleading about its purpose and that's why I don't like it much.

Thanks again for the help.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

This version looks much better.  Nice job. :)  So, about point 2.3...   To elaborate on this, the new lines in microdom.py:

```python
    if isinstance(self.parentNode, Document) or self.parentNode is None:
        if self.namespace is not None and namespace == '':
            namespace = self.namespace
```

are incompletely tested.  If I change them to this:

```python
    if self.namespace is not None and namespace == '':
        namespace = self.namespace
```

then all of the twisted.web unit tests continue to pass.  trial --coverage is a handy tool, but it doesn't _quite_ report the entire story.  The way I determined the first line could be removed is by removing it and then running the tests. :)  Since they all pass, the line is "untested" - which is a bit inaccurate, since lots of tests do run it, but they all run it in the same way - they all test the "positive" path through it, and none of them tests the "negative" path through it.  If any of the tests relied on that if test evaluating to false and preventing the next two lines from running, deleting that line would have caused them to fail.

It's a bit too late at night for me to really dig into the details of why the line is correct (or why it might be wrong).  Hopefully the above explanation gives you a better idea of how to expand the test coverage for it, though.

And a hint about corner cases - writing unit tests is a really awesome way to track them all down. :)

**Comment by twisted-trac** [comment]
|<img alt="shywolf9982's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|shywolf9982 commented|
|-|-|

You were right :).

Took me some time, and the test could be simplified (there's probably a spurious tag midway into it) but now removing that line causes a failure.

My only concern is that the specification isn't so clear, so the actual behaviour is based on my 
_interpretation_ of the specification (interpolating its parts).

It might not be the correct interpretation, however.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

This should be unassigned to indicate that anyone may review it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

The patch seems pretty good now, but I still think microdom is doing something wrong.  Unfortunately I think the only way I'll be able to figure out exactly what is by trying to fix it, so I'm going to try to do that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [26383]) Branching to 'microdom-ns-3672'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I think this is good now.  Maybe.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

I think this change is pretty good.  Thanks for all the comments.

    1. Extremely Minor Nits:
        1. on line `microdom.py:660`, there's still inconsistent quoting style: '"' for <, "'" for >.  it would be nice to fix that before merging.
        1. it would similarly be nice to add a blank line before and after `writexml`.
    1. maybe `writexml` shouldn't be recursive.  If `nsprefixes` and `namespace` are purely implementation details, maybe `writexml` should initialize them and call `_writexml` which documents the parameters without the caveat, since the underscore presumes implementation-detail-hood; `writexml`'s documentation could then simply say that the parameters are deprecated and their semantics are not useful.  (and, of course, emit the appropriate deprecation warning if they're passed).

Address these to your satisfaction - moving point 2 into a separate ticket if you would rather do it later - and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [26425]) Use consistent string syntax; add appropriate whitespace before and after writexml method

refs [#3672](https://github.com/twisted/twisted/issues/3672)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set status to `closed`|
|-|-|

(In [26428]) Merge microdom-ns-3672 

Author: shywolf9982, exarkun
Reviewer: exarkun, glyph
Fixes: [#3672](https://github.com/twisted/twisted/issues/3672)

Fix microdom's serialization of elements with respect to namespaces.  In
particular, emit the correct closing tag when the corresponding open tag is
qualified with a namespace prefix (previously the namespace prefix would be
omitted from the closing tag) and correctly track which namespace is active
as the default.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
